File: tests/sdk/nodejs/test/macie-findings.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { newS3, requireReady, signedJsonTarget, uniquePrefix } from "../lib/helpers.mjs";

test("Macie EnableMacie InjectFindings List/GetFindings", async (t) => {
  if (!(await requireReady(t))) return;
  if (process.env.NOCTAXRIS_MACIE_INJECT !== "1") {
    t.skip(
      "set NOCTAXRIS_MACIE_INJECT=1 on the API process for lab InjectFindings",
    );
    return;
  }

  const s3 = newS3();
  const prefix = uniquePrefix();
  const bucket = `${prefix}-macie`.toLowerCase();
  const key = "pii.csv";
  const body = Buffer.from("ssn 987-65-4321\n");

  const enable = await signedJsonTarget("macie2", "Macie2.EnableMacie", {});
  assert.equal(enable.status, 200, enable.body);

  await s3.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(async () => {
    try {
      await s3.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
    } catch {
      /* ignore */
    }
    try {
      await s3.send(new DeleteBucketCommand({ Bucket: bucket }));
    } catch {
      /* ignore */
    }
  });
  await s3.send(
    new PutObjectCommand({ Bucket: bucket, Key: key, Body: body }),
  );

  const inj = await signedJsonTarget(
    "macie2",
    "NoctaxrisMacie.InjectFindings",
    {
      S3Objects: [{ Bucket: bucket, Key: key }],
    },
  );
  assert.equal(inj.status, 200, inj.body);
  const findingIds = inj.json?.findingIds || [];
  assert.ok(findingIds.length >= 1, `InjectFindings findingIds empty: ${inj.body}`);

  const listed = await signedJsonTarget("macie2", "Macie2.ListFindings", {});
  assert.equal(listed.status, 200, listed.body);
  const listedIds = listed.json?.findingIds || [];
  assert.ok(
    listedIds.includes(findingIds[0]),
    `ListFindings missing ${findingIds[0]}: ${listed.body}`,
  );

  const got = await signedJsonTarget("macie2", "Macie2.GetFindings", {
    findingIds,
  });
  assert.equal(got.status, 200, got.body);
  assert.ok(
    got.body.includes("SensitiveData"),
    `GetFindings missing SensitiveData: ${got.body}`,
  );
  assert.ok(
    (got.json?.findings || []).length >= 1,
    `GetFindings findings empty: ${got.body}`,
  );
});

File: tests/sdk/go/macie_findings_test.go
package sdk_test

import (
	"bytes"
	"context"
	"os"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func TestMacieEnableInjectListGetFindings(t *testing.T) {
	requireReady(t)
	if os.Getenv("NOCTAXRIS_MACIE_INJECT") != "1" {
		t.Skip("set NOCTAXRIS_MACIE_INJECT=1 on the API process for lab InjectFindings")
	}
	cfg := loadAWSConfig(t)
	s3c := newS3(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)
	bucket := strings.ToLower(prefix + "-macie")
	key := "pii.csv"
	body := []byte("ssn 987-65-4321\n")

	enableStatus, enableBody, _ := signedJSONTarget(t, "macie2", "Macie2.EnableMacie", map[string]any{})
	if enableStatus != 200 {
		t.Fatalf("EnableMacie status=%d body=%s", enableStatus, enableBody)
	}

	if _, err := s3c.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)}); err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		_, _ = s3c.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
		_, _ = s3c.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})
	if _, err := s3c.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket), Key: aws.String(key), Body: bytes.NewReader(body),
	}); err != nil {
		t.Fatalf("PutObject: %v", err)
	}

	injStatus, injBody, injParsed := signedJSONTarget(t, "macie2", "NoctaxrisMacie.InjectFindings", map[string]any{
		"S3Objects": []map[string]any{
			{"Bucket": bucket, "Key": key},
		},
	})
	if injStatus != 200 {
		t.Fatalf("InjectFindings status=%d body=%s", injStatus, injBody)
	}
	rawIDs, _ := injParsed["findingIds"].([]any)
	if len(rawIDs) < 1 {
		t.Fatalf("InjectFindings findingIds empty: %s", injBody)
	}

	listStatus, listBody, listParsed := signedJSONTarget(t, "macie2", "Macie2.ListFindings", map[string]any{})
	if listStatus != 200 {
		t.Fatalf("ListFindings status=%d body=%s", listStatus, listBody)
	}
	listedIDs, _ := listParsed["findingIds"].([]any)
	if !jsonAnyContains(listedIDs, rawIDs[0]) {
		t.Fatalf("ListFindings missing injected id %v: %s", rawIDs[0], listBody)
	}

	getStatus, getBody, getParsed := signedJSONTarget(t, "macie2", "Macie2.GetFindings", map[string]any{
		"findingIds": rawIDs,
	})
	if getStatus != 200 {
		t.Fatalf("GetFindings status=%d body=%s", getStatus, getBody)
	}
	if !strings.Contains(string(getBody), "SensitiveData") {
		t.Fatalf("GetFindings missing SensitiveData: %s", getBody)
	}
	findings, _ := getParsed["findings"].([]any)
	if len(findings) < 1 {
		t.Fatalf("GetFindings findings empty: %s", getBody)
	}
}

func jsonAnyContains(haystack []any, needle any) bool {
	want := stringifyAny(needle)
	for _, item := range haystack {
		if stringifyAny(item) == want {
			return true
		}
	}
	return false
}

func stringifyAny(v any) string {
	switch t := v.(type) {
	case string:
		return t
	default:
		return ""
	}
}
